### DISCUSSION TITLE
Naming conventions in Wheels

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1366] [Q&A]
Just wanting to know what the naming conventions are in CFWheels. As I recall seeing:
- Controller CFC files are named as plural and in PascalCase. Ex. Users.cfc.
- Method names should be in camelCase.
- View CFM files to match the action methods in the controller are placed in a folder to match the controller name, but all lower case. CFM file is also all lower case. Ex: /views/users/new.cfm
- Model CFC files are named as singular?  Camel case?  Name to match the db table name that is of the same name (doesn't have to be) but in plural form. Ex. /model/Author.cfc maps to table "authors" in db.

I can't say the above is consistent IMHO. What is the naming convention best practices. Please correct/advise on this as I'm trying to refactor an app that didn't follow any of this and I don't want to refactor it twice to get it right.

**Comment by chapmandu** [comment]
All correct! 👍

**Comment by cfcoder** [comment]
Does the model name have to match the controller name by convention? If controller is Users, would the model be Users or User?

Why is the controller name in the plural form and the model name singular? I’ve seen tutorials that have controller names that are singular, like "User".

**Reply by chapmandu** [reply]
The model doesn't have to match the controller. You can name your controllers anything you want. 

For better or worse, the conventions are based on Rails.

**Comment by cfcoder** [comment]
Then how does CFWheels know where to find the model for a given controller? Are the only referenced in the new model() calls? I read that models are optional so there must be some default mapping going on.
On Dec 7, 2023 at 5:08 PM -0800, Adam Chapman ***@***.***>, wrote:
	The model doesn't have to match the controller. You can name your controllers anything you want.
For better or worse, the conventions are based on Rails.

**Reply by chapmandu** [reply]
There is no convention for binding a model to a controller. Perhaps the examples you've seen imply this exists, but it doesn't. You must explicitly reference models from your controller. 

You CAN omit a model file and use wheels convention to assume your table name, but I don't employ this feature.

**Comment by bpamiri** [comment]
In the early days of CFWheels there was a strong correlation between the URLs of an app to the Controllers and Actions. Much of this has gone away with the advent of the new router circa CFWheels 2.0 and the use of resource based routing. Although you can still achieve the same 1.x level of URL to controller/action linking with the use of the wildcard route definition.

Having said that, it is reasonable to assume in a purely CRUD based admin app, one might find a one to one mapping between a controller and a model. But as routes have diverged from the simple "/controller/action" nomenclature it is reasonable to imagine a single controller accessing multiple models. Think for instance of a Dashboard where you display a bunch of different data to your user. The route that displays the Dashboard may be /dashboard and the action may be defined in the index action of the Dashboard controller. But that index action may pull in data from multiple models like "invoice", "message" or other elements you may want to display on the page.

This give you, the developer, more freedom to structure your routes, controllers, and models in whatever way makes the most sense for your particular app.
